### PR TITLE
Fix critical bug with attribute reordering by destructive Array.Reverse() on memoized array; breaking changes of interface to IReadOnlyList

### PR DESCRIPTION
**Pull request by SergeiPavlov**

### CODE REVIEW

File: Orm/Xtensive.Orm/Core/ReversedReadOnlyList.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// See the License.txt file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xtensive.Core
+{
+  public struct ReversedReadOnlyList<T> : IReadOnlyList<T>
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Core/ReversedReadOnlyList.cs:12]: `readonly struct` ?

Comment (AlexUstinov) [reply Orm/Xtensive.Orm/Core/ReversedReadOnlyList.cs:12]: Do we really want to expose it public?

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Core/ReversedReadOnlyList.cs]: made `internal readonly struct`

File: Orm/Xtensive.Orm/Orm/Building/Builders/ModelDefBuilder.cs [hunk 2]
```diff
@@ -286,7 +286,7 @@ public HierarchyDef DefineHierarchy(TypeDef typeDef, HierarchyRootAttribute attr
     public FieldDef DefineField(PropertyInfo propertyInfo) =>
       DefineField(propertyInfo, GetFieldAttributes<FieldAttribute>(propertyInfo));
 
-    public FieldDef DefineField(PropertyInfo propertyInfo, FieldAttribute[] fieldAttributes)
+    public FieldDef DefineField(PropertyInfo propertyInfo, ReversedReadOnlyList<FieldAttribute> fieldAttributes)
```
Comment (AlexUstinov) [inline Orm/Xtensive.Orm/Orm/Building/Builders/ModelDefBuilder.cs:289]: Should `ReversedReadOnlyList<FieldAttribute> fieldAttributes` be declared as `in` parameter?

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/Builders/ModelDefBuilder.cs]: fixed

**Review by AlexUstinov — APPROVED** [review]

**Pull request merged**